### PR TITLE
K-fold cross validation

### PR DESCRIPTION
**Pull request by lperozzi**
I used the @CannedGeo solution and implemented the @LukasMosser [K-fold cross validation](https://gist.github.com/LukasMosser/cd645bad2bdbbb419098ac3ea363f2b3) to show how different F1 score is between each well. Any comments is welcome

### CODE REVIEW

**Comment by kwinkunks** [comment]
Thank you for doing this Lorenzo, I'll give it a look right away.

Sorry for the slow reaction time!

**Comment by kwinkunks** [comment]
I notice that the cross validation stuff has changed a little in `scikit-learn` 0.18. There are a couple of k-fold CV classes in there... has anyone played with them? 

Docs >> http://scikit-learn.org/stable/tutorial/statistical_inference/model_selection.html#cross-validation-generators

(I will add this to Issues, rather than starting a discussion here... please go there)

**Comment by CannedGeo** [comment]
@lperozzi and @LukasMosser this is great and very concise, thank you. I think the idea of an average F1 score makes sense. But maybe if this is used, we exclude the F9 well? Since this only includes one facies and may weight the output average?

**Comment by LukasMosser** [comment]
Having read a bit about cross-validation in the deep learning book from MIT, I've realized a couple things:

Our assumption when training our data is that each data point behaves independently from all the others. Essentially that means that there is no reason why we can't aggregate all our data in a a big blob, split it up in training, validation (k-fold) and a test set (Shankle). 

I believe the important part here is that Shankle should really be a test set, and when training our data we shouldn't be actively testing against performance on Shankle rather than performance on our k-fold validation. 

On the other hand,  I believe the assumption that our data points can be considered independent, is not correct as we are dealing with a continuous dataset where the occurence of the facies and their spatial relationships i.e. sand 1 is followed by sand 2 etc. can't be neglected. Possibly a reason why all our approaches don't achieve very good results. 

If we perform cross-validation and keep Shankle as a test dataset, then this also means that Shankle must be a _representative_ dataset similar to one of the folds from our cross validation. Otherwise we may be under or overestimating the performance of our trained classifiers. 

So to summarise:
- Using well based k-fold cross validation would make sense, when evaluating classification of the data as ordered sets where spatial relationships matter.
- Shankle has to be a representative dataset that allows us to evaluate our classifiers.
- Training and hyperparameter optimisation should not be performed against Shankle or a representative test dataset at any time.
- With any of the classifiers such as SVC or random forest spatial relationships aren't taken into account, therefore there is no point in keeping data separate in wells. We may as well keep it in a blob.
  It is interesting to keep them separate nevertheless.
- The test set should either be a representative well, with representative occurences of facies allowing spatial relationships to be taken into account, or all our data should be put into a bucket, take away a secret test set, and provide the rest for k-fold cross validation. On submission, the approaches are run against the test set and results posted by the judges.

What do you think guys? @lperozzi @kwinkunks @CannedGeo
